### PR TITLE
[Discv5] Generic queries

### PR DESCRIPTION
**Pull request by pawanjay176**
## Description

Refactor to have more generic queries. This mostly follows from the query structure in `protocols/kad`. 
Currently, we only have one query type: iterative FINDNODE queries. 
This would be useful for adding additional query logic like the predicate based ENR query.

With these changes, behaviour now interacts with a `QueryPool` which drives queries with different logic.

Edit: Also adds a timeout to the `FindNodeQuery`. If a peer doesn't reply to a query within the designated timeout (currently set to the Discv5Config `request_timeout` in behaviour), the peer's `QueryPeerState` is marked as `Failed`.

### CODE REVIEW

**Comment by pawanjay176** [comment]
@AgeManning this is ready for review :)

File: misc/discv5/src/config.rs [hunk 1]
```diff
@@ -7,6 +7,11 @@ pub struct Discv5Config {
     /// The request timeout for each UDP request. Default: 4 seconds.
     pub request_timeout: Duration,
 
+    /// The query timeout after which a `QueryPeer` is marked unresponsive.
+    /// Unresponsive peers don't count towards the parallelism limits for a query.
+    /// Hence, we may potentially end up making more requests to good peers.
+    pub query_timeout: Duration,
```
Comment (AgeManning) [inline misc/discv5/src/config.rs]: We should add a function to the Discv5ConfigBuilder to allow configuration of this new parameter

File: misc/discv5/src/query_pool/peers/closest.rs [hunk 2]
```diff
@@ -62,65 +64,34 @@ pub struct QueryConfig {
...
+    ///
+    /// If a successful result is not reported for a peer within this timeout
+    /// window, the iterator considers the peer unresponsive and will not wait for
+    /// the peer when evaluating the termination conditions, until and unless a
+    /// result is delivered. Defaults to `10` seconds.
+    pub peer_timeout: Duration,
 }
 
-impl Default for QueryConfig {
+impl Default for FindNodeQueryConfig {
```
Comment (AgeManning) [inline misc/discv5/src/query_pool/peers/closest.rs]: I think potentially, we remove the `Default` impl and instead create a `new_from_config(config: &Discv5Config)` and we already set the defaults in the config and should make the code a bit nicer.

This way we won't have mismatched default values

File: misc/discv5/src/query_pool/peers/closest.rs [hunk 3]
```diff
@@ -215,6 +176,24 @@ where
                         peer.state = QueryPeerState::PendingIteration;
                     }
                 }
+                QueryPeerState::Unresponsive => {
+                    let peer = e.get_mut();
+                    peer.peers_returned += num_closest;
```
Comment (AgeManning) [inline misc/discv5/src/query_pool/peers/closest.rs]: I think we should set this to be:
```rust
peer.peers_returned += closer_peers.len();
```

This broke some of our tests last time. We may need to investigate why it breaks those tests, as I believe this is the correct logic.

File: misc/discv5/src/query_pool/peers/closest.rs [hunk 4]
```diff
@@ -215,6 +176,24 @@ where
...
+                    if peer.peers_returned >= self.config.num_results {
+                        peer.state = QueryPeerState::Succeeded;
+                    } else if self.iterations == peer.iteration {
+                        if peer.peers_returned > 0 {
+                            // mark the peer as succeeded
+                            peer.state = QueryPeerState::Succeeded;
+                        } else {
+                            peer.state = QueryPeerState::Failed; // didn't return any peers
```
Comment (AgeManning) [inline misc/discv5/src/query_pool/peers/closest.rs]: I thought we were removing this?

Comment (AgeManning) [reply misc/discv5/src/query_pool/peers/closest.rs]: I think this logic is sane however

File: misc/discv5/src/query_pool/peers/closest.rs (hunk 3 above)
Comment (pawanjay176) [reply misc/discv5/src/query_pool/peers/closest.rs]: Will open another PR to fix this to keep it logically separate.

File: misc/discv5/src/query_pool/peers/closest.rs (hunk 4 above)
Comment (pawanjay176) [reply misc/discv5/src/query_pool/peers/closest.rs]: ^ same as above

**Pull request merged**